tools/git-change-exec/actions.go

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"os"
+	"os/exec"
+	"reflect"
+)
+
+type action interface {
+	match(path string) bool
+	do() error
+}
+
+func id(i any) string {
+	ty := reflect.TypeOf(i)
+	if ty.Name() == "" {
+		ty = reflect.TypeOf(i).Elem()
+	}
+	return ty.Name()
+}
+
+func execCmdWithDefaults(name string, args ...string) *exec.Cmd {
+	cmd := exec.Command(name, args...)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	return cmd
+}
+
+// Do not forget to add your Action HERE
+var actions = map[string][]action{
+	"test": []action{
+		pillarTestAction{},
+		gitChangeExecTest{},
+		getDepsTestAction{},
+		bpftraceCompilerExecTest{},
```

Comment:
Seems like all of these are of the form
if match on directory prefix, then run this command

Do we need to put that as golang code or could we e.g., have that in some separate file? E.g., an actions_test file will be used for test, actions_lint file for lint, and actions_foo for a future foo target?
Or have the action filename be an argument to the tool instead of passing e.g., "test" as the argument?

Reply:
Currently it is in different files `actions-test.go` and `actions-lint.go`, but unfortunately it is not possible with reflection to read which structs are there that implement the `action` interface.

As an experiment I tried to use a golang interpreter: https://github.com/lf-edge/eve/commit/66803f01ee234a15d264656508c71c1dfeb94654#diff-5bf19742ee6060707f406c1625c23226124e161d2049d12a80a8a61ef076ace8R72
But then I thought, what if there are dependencies between actions (like f.e. for linting, one would like first to run the spdx-fixer and then `go fmt`), so I would have to implement a way to express dependencies.

Reply:
But I don't see why we need the generality of golang to run the actions, since the actions do simple matches on paths and then exec code. And it would be a lot more modular and easier to understand if we don't have to put various make invocations in the golang code but keep that in separate files IMHO.

Reply:
> I don't see why we need the generality of golang to run the actions

F.e. for the `lint` action I need it, because it has to edit files and add the copyright statement.
Of course I could do it that `git-change-exec` invokes another program to do so, but then I have to add:
* finding the path to the other program
* serialization
* deserialization including error(-string) parsing
* eventually implementing a way to save state between `match` and `do` and handle a proper cleanup

(the first point makes it harder to develop this tool, because I test it on another clone of the eve repository)

> more modular and easier to understand

I changed the code so that it runs `make test` for the tests, so the `Makefile` is the abstraction, but I fear that in the future the actions become more complex, so only `type action interface` will be an abstraction.

>  since the actions do simple matches on paths 

That was just, because I did not add yet to ignore `README`s and markdown files; also `pkg/pillar/agentlog/cmd` does not need to trigger the go tests. Now it is there.
The `lint` action has to save the matches in an array to be able to fix them in the `do` method.

Generally speaking I think `match` can become a lot more complex in the future once f.e. we only run the tests based on code coverage.

> but keep that in separate files

What would these files look like?

---

Suggestions from my side:

1. put every action in it's own go-file

2. enhance `dry-run` to show the match result of every file

3. add a `doDryRun()` to the `action` interface where every action has to describe what it plans to do